File: Java/src/session2/td/heritageMultiple/NatParIntRecursif.java
package session2.td.heritageMultiple;

import session1.td.FabriqueNaturels;
import session1.td.Nat;

public class NatParIntRecursif extends IntPositif implements AlgebreNatRecursif {

    public static final FabriqueNaturels<Nat> FAB = new NatParIntRecursif(0);

    public NatParIntRecursif(int val) {
        super(val);
    }

    @Override
    public boolean estEgal(Object o) {
        if(!(o instanceof Nat))
            return false;
        Nat x = (Nat)o;
        if(x.estNul())
            return false;
        return this.predecesseur().equals(x.predecesseur());
    }



    @Override
    public String toString() {
        return  "" + this.val();
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof Nat))
            return false;
        Nat x = (Nat)obj;
        return this.val() == x.val();
    }

    @Override
    public Nat creerNatAvecValeur(int val) {
        return new NatParIntRecursif(val);
    }


}

File: TypeScript/src/session2.td/heritageMultiple/NatParIntRecursif.ts
import {Nat} from "../../session1.td/Nat";
import {FabriqueNaturels} from "../../session1.td/FabriqueNaturels";
import {IntPositif} from "./IntPositif";
import {AlgebreNatRecursif, AlgebreNatRecursif_default} from "./AlgebreNatRecursif";
import {heriterInterfacesConcretes} from "./heritageInterfacesconcretes";


export class NatParIntRecursif extends IntPositif implements AlgebreNatRecursif {

    public static FAB : FabriqueNaturels<Nat> = new NatParIntRecursif(0);

    public constructor(val : number) {
        super(val);
    }

    public creerNatAvecValeur(val : number) : Nat {
        return new NatParIntRecursif(val);
    }

    public toString() : string {
        return String(this.val());
    }

    public equals(obj : Nat) : boolean {
        return this.val() == obj.val();
    }

    div(x: Nat): Nat {
        return undefined;
    }

    modulo(x: Nat): Nat {
        return undefined;
    }

    produit(x: Nat): Nat {
        return undefined;
    }

    somme(x: Nat): Nat {
        return undefined;
    }

    un(): Nat {
        return undefined;
    }

    zero(): Nat {
        return undefined;
    }
}

heriterInterfacesConcretes(NatParIntRecursif, [IntPositif, AlgebreNatRecursif_default]);